Feat/#5 reservation 예매 기능

kwj0605 opened the pull request:
### PR 타입
- [x] 기능 추가
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
### 반영 브랜치
- feature/이슈번호/(create · read · update · delete) - 도메인명 -> main
### 이슈
- #5 
### Description
- 예약 기능 추가 -> 리팩토링
### To Reviewers
- 권한체크 부분의 if문 로직이 맞는지 확인 부탁드립니당

kwj0605 committed:
feat "공통 ResponseDto, Exception 생성" #5

kwj0605 committed:
feat "공통 ResponseDto, Exception 생성" #5

kwj0605 committed:
setup jpa의존성 추가 #5

kwj0605 committed:
feat 'Base' 생성 #5

kwj0605 committed:
feat 'ResponseDataDto' 수정 #5
httpStatus int로 수정

kwj0605 committed:
feat Entity 생성 #5

kwj0605 committed:
feat PageableUtil 생성 #5
기본 정렬 - createdAt, DESC
페이지 사이즈 - 10(default), 30, 50

kwj0605 committed:
feat 예약 내역 전체 조회 #5

kwj0605 committed:
refactor 주석, @transactional 추가 #5

kwj0605 committed:
setup build.gradle 'openfeign' 추가 #5

kwj0605 committed:
feat Feign Client 추가 #5
공연, 타임테이블, 좌석, 사용자

kwj0605 committed:
refactor 변수 이름 변경 #5

kwj0605 committed:
refactor 변수 이름 변경, 불필요한 변수 삭제 #5

kwj0605 committed:
setup Lombok 의존성 추가 #5

kwj0605 committed:
refactor 클래스 명 변경  #5

kwj0605 committed:
fix 예매 내역 조회 로직 수정 #5
feign client로 필요한 값 가져오기
(user, seat, timetable, performance)

kwj0605 committed:
setup git ignore 수정 #5
(windows,java,intellij,gradle,macos)

kwj0605 committed:
setup postgresql 의존성 추가 #5

kwj0605 committed:
git merge #5
 git merge to dev

kwj0605 committed:
fix feign client 중복 오류 수정 #5

kwj0605 committed:
fix 메서드 명 수정 #5

kwj0605 committed:
feat 예약 내역 단일 조회 #5

kwj0605 committed:
feat 내 예약 내역 조회 #5

kwj0605 committed:
setup build.gradle 수정 #5
dependencyManagement가 dependencies보다 위에 있도록 수정

kwj0605 committed:
feat 예약 내역 검색 #5

kwj0605 committed:
setup 'application.yml' application 이름 변경 #5

kwj0605 committed:
feat 예약 취소(삭제) #5

kwj0605 committed:
refactor 예약 취소(삭제) #5
- 권한체크 추가
- pageableUtil 컴포넌트 주입으로 수정
- 예약 내역 검색 USER일 경우 로그인 유저가 작성한 것만 조회하도록 수정

yshong1998 commented on com.culture-ticket.client.reservation-payment/src/main/java/com/culture_ticket/client/reservation_payment/application/service/ReservationService.java:
```diff
@@ -30,4 +54,179 @@ public void createReservation(ReservationRequestDto request) {
+     */
+    @Transactional(readOnly = true)
+    public Page<ReservationResponseDto> getReservations(String role, Pageable pageable) {
+        if (!role.equals("ADMIN")) {
```
권한을 검증하는 로직이 아무래도 메서드마다 반복적으로 많이 사용되다 보니, 저는 아래처럼 RoleValidator라는 클래스를 만들어서 static 메서드로 검증 메서드를 만들어서 호출하는 편이긴 합니다!
혹시 아래처럼 권한 관리 클래스를 common이나 util 부분에 추가해서 관리하는 건 어떠신가요?

public class RoleValidator {

  public static void validateIsMaster(String role) {
    if (!role.equals("MASTER")) {
      throw new IllegalArgumentException("인가되지 않은 사용자입니다.");
    }
  }

  public static void validateIsHubManager(String role) {
    if (!(role.equals("HUB_MANAGER") || role.equals("MASTER"))) {
      throw new IllegalArgumentException("인가되지 않은 사용자입니다.");
    }
  }

  public static void validateIsHubDeliveryManager(String role) {
    if (!(role.equals("HUB_TO_HUB_DELIVERY") || role.equals("MASTER"))) {
      throw new IllegalArgumentException("인가되지 않은 사용자입니다.");
    }
  }
}

kwj0605 replied:
오! 좋은 방법인 거 같아요. 안그래도 권한체크들이 많아질 경우 중복되는 게 많아질 거 같아서 고민 중이였는데..
utilClass로 따로 빼서하는 거 좋은 거 같습니당!

SEJ123 replied:
private void validateRoleADMIN(String role) {
        if (!role.equals("ADMIN")) {
            throw new CustomException(ErrorType.FORBIDDEN);
        }
    }
    저는 서비스 내에 권한 검증하는 메소드를 추가해서 진행했는데, 클래스로 따로 만드는 이유가 있을까요??

yshong1998 replied:
중복코드가 많이 발생해서 가독성 측면에서 좋지 않아서 빼는 느낌이 제일 강하긴 합니다! 클래스를 빼는 것으로 얻을 수 있는 이점은 크게 아래 둘 정도일 것 같아요.
1. 매 코드마다 throw new CustomException("메세지");를 추가하기 위해 3줄씩 늘어나는 코드를 1줄로 줄일 수 있습니다.
2. 코드에 .equlas("value")와 같은 로직이 밖으로 드러나 있기보다는 로직은 따로 만든 클래스에 구현하고 서비스 레이어에서는 메서드가 validateRoleAdmin과 같은 네이밍으로 되어 있으면 '권한을 검증하는 로직'이 수행되고 있음을 바로 이름에서 알 수 있습니다. 만약 .equals(value);를 서비스 레이어에 그대로 두면 하면 그 코드를 해석하고 '아 권한 검증 로직이구나'를 파악해야 해서 가독성 측면에서 더 불리합니다.

yshong1998 commented on com.culture-ticket.client.reservation-payment/src/main/java/com/culture_ticket/client/reservation_payment/application/service/ReservationService.java:
```diff
@@ -30,4 +54,179 @@ public void createReservation(ReservationRequestDto request) {
+        reservation.deleted(username);
+    }
+
+    private static void isReservationInPage(Page<Reservation> reservationPage) {
```
혹시 이 메서드는 어떤 용도로 활용되는 걸까요? 로직상 안에 아무 것도 없으면 예외가 발생하는 것 같은데
예매를 한 번도 한 적이 없다면 아무 것도 없을 수도 있지 않을까 해서요!

kwj0605 replied:
아하 그러네용 없으면 예약 내역이 비어있는 상태로 조회할 수 있군요.. 수정하겠습니다!

SEJ123 commented:
Controller - @RequestMapping("api/v1/reservation") 엔드 포인트를 복수형으로 수정해주시면 될 거 같습니다!

yshong1998 commented on com.culture-ticket.client.reservation-payment/src/main/java/com/culture_ticket/client/reservation_payment/application/service/ReservationService.java:
```diff
@@ -30,4 +54,179 @@ public void createReservation(ReservationRequestDto request) {
+
+        // 결제 삭제
+        Payment payment = reservation.getPayment();
+        payment.deleted(username);
```
예약을 취소해서 결제가 사라진다면, payment의 경우 환불이 되어야 하는데 저희는 로직 상 비용 없이도 결제가 되는 중이긴 해서 환불 로직이 구체적으로는 없는 게 맞는 것 같아요.
그래서 생각한 건데 예약 취소 시 반환값으로 환불액(refundAmount, refundPrice등) 필드를 추가하는 건 어떨까요? 값은 결제가 totalPrice를 필드로 갖고 있어서 이 값을 그냥 그대로 넣어서 이름만 바꿔서 반환해도 괜찮을 것 같습니다!

kwj0605 replied:
오.. 좋은 생각인 거 같아요! 그렇게 처리하면 나중에 환불관련 서비스를 추가할 때 더 수월하겠네요.

yshong1998 commented on com.culture-ticket.client.reservation-payment/src/main/java/com/culture_ticket/client/reservation_payment/common/util/PageableUtil.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class PageableUtil {
```
오... 이렇게도 사용을 하는군요 간단하게 사용할 때는 @PageableDefault로도 충분하지만
이렇게 10,30,50처럼 가능한 값이 미리 정해져 있는 경우 이렇게 유틸 클래스 만드는 것 좋은 것 같습니다!

kwj0605 replied:
10, 30, 50 이외의 값이 안들어 오도록 막으려고 생각해봤어요 :)

yshong1998 commented on com.culture-ticket.client.reservation-payment/src/main/java/com/culture_ticket/client/reservation_payment/infrastructure/client/PerformanceClient.java:
어떤 네이밍과 구조가 맞는 건지 확실한 근거는 없는 부분이기는 합니다! 의견으로 들어주세요.
개인적으로는 feignClient로 받아오게 될 객체는 영속 상태의 객체가 아닌 Dto로서 값만 복사된 객체이기 때문에 저는 feignClient에서 받아올 객체의 이름에도 뒤에 Dto를 붙이는 편이고 Domain 레이어가 아닌 Infrastructure 하위에 dto를 하나 더 둬서 그곳에 두는 것이 좋다는 생각입니다.
혹시 어떻게 생각하실까요?

kwj0605 replied:
저도 처음에 예석님 생각처럼 Entity를 만들어 가져오는 것보다 DTO로 필요한 값만 가져와서 사용하는 게 더 안전하지 않을까? 라고 생각했다가.. 강의에서 Entity로 만들어 사용하길래 무작정 따라 했던 거 같습니다. 다른 분들에게도 여쭤보니 DTO로 하는 분이 더 많은 거 같습니다! 수정하겠습니다.

yshong1998 commented on com.culture-ticket.client.reservation-payment/src/main/java/com/culture_ticket/client/reservation_payment/presentation/controller/ReservationController.java:
```diff
@@ -0,0 +1,166 @@
+    @GetMapping
+    public ResponseEntity<ResponseDataDto<Page<ReservationResponseDto>>> getReservations(
+        @RequestHeader(value = "X-Role", required = true) String role,
+        @RequestParam(defaultValue = "0") int page,
```
@PageableDefault를 쓰면 한 번에 받아올 수 있습니다!

yshong1998 reviewed (approved):
리뷰 남겼습니다! 문제가 되는 부분은 없어서 approve로 해 두겠습니다
제안 사항은 그래도 있기는 해서 확인 해 보시고 답글 남겨주시면 좋을 것 같습니다

kwj0605 committed:
refactor 검색 조건 수정 #5
검색조건 - 1달전, 3달전, 6달전, 1년전

kwj0605 committed:
refactor isReservationInPage 삭제 #5

kwj0605 committed:
refactor 예약 내역 리팩토링 #5
- isReservationInPage 메서드 삭제
(이유: 데이터가 없으면 에러가 아니라 없는 페이지를 보여줘야함)
- RoleValidator 생성
(이유: 반복되는 권함 검증 로직을 처리)
- 검색 조건 영어로 수정

kwj0605 committed:
refactor 예약 내역 리팩토링 #5
- 내 예약 내역 조회 시 isDeleted가 false만 조회
- 예약 취소 시 반환되는 환불액 dto로 감싸기

kwj0605 committed:
refactor 예약 내역 리팩토링 #5
- Feign Client로 받아오는 값들 Dto로 변경(Entity -> Dto)

kwj0605 committed:
refactor RepositoryCustom 위치 변경 #5

kwj0605 committed:
Merge branch 'dev' of https://github.com/Sparta-Triple/CultureTicket into feat/#5-reservation

kwj0605 committed:
refactor 자료형 변경 #5
- 타임테이블에서 feign client로 가져오는 공연날짜의 자료형 통일

kwj0605 committed:
refactor api 복수로 수정 #5

kwj0605 committed:
refactor feign client 테스트 후 수정 #5
- Json 역직렬화 해결 -> FeignClientResponseDataDto
- X-User-Role 이름 수정

kwj0605 committed:
refactor 예약 내역 테스트 후 수정 #5
- 유저 role일 경우만 데이터 소유주 검사

kwj0605 committed:
refactor 결제 수정 #29
- 결제 생성할 때 좌석 결제 내역 안생기던 오류 수정
- 결제 생성 시 결제 테이블에 총가격 들어가도록 수정

kwj0605 committed:
Merge branch 'dev' of https://github.com/Sparta-Triple/CultureTicket into feat/#5-reservation

kwj0605 committed:
git merge #5

kwj0605 committed:
fix 결제 생성 시 티켓 생성 #5

kwj0605 committed:
Merge branch 'dev' of https://github.com/Sparta-Triple/CultureTicket into feat/#5-reservation

kwj0605 committed:
fix 티켓 생성 로직 수정 #5

kwj0605 committed:
fix 예약 삭제 시 티켓 삭제 로직 추가 #5

kwj0605 committed:
fix 테스트 코드 주석처리 #5

kwj0605 committed:
fix feign client 수정 #5

kwj0605 committed:
Merge branch 'dev' of https://github.com/Sparta-Triple/CultureTicket into feat/#5-reservation

kwj0605 committed:
fix 테스트 및 수정 #5
- Feign client에서 patch메서드 사용이 안됨 -> Put으로 수정
- 좌석 단일 조회 추가
- 타임테이블 단일 조회 추가

kwj0605 committed:
fix 테스트 및 수정 #5
- user 정보 조회 권한 해제
